### PR TITLE
feat(router): add v2 api for /fingerprint

### PR DESCRIPTION
**Pull request by ShankarSinghC**
add v2 api for /fingerprint

### CODE REVIEW

**Review by Chethan-rao — APPROVED** [review]

File: src/storage/db.rs [hunk 1]
```diff
@@ -345,12 +345,12 @@ impl super::FingerprintInterface for Storage {
...
-        card: types::CardNumber,
-        hash_key: Secret<String>,
+        data: Secret<String>,
+        key: Secret<String>,
     ) -> Result<types::Fingerprint, ContainerError<Self::Error>> {
-        let algo = HmacSha512::<1>::new(hash_key.expose().into_bytes().into());
+        let algo = HmacSha512::<1>::new(key.expose().into_bytes().into());
 
-        let fingerprint_hash = algo.encode(card.into_bytes())?;
+        let fingerprint_hash = algo.encode(data.expose().into_bytes())?;
```
Comment (NishantJoshi00) [inline src/storage/db.rs:353]: can we change the definition of encode

```diff
- impl<const N: usize> Encode<Vec<u8>, Vec<u8>> for HmacSha512<N> {
+ impl<const N: usize> Encode<Secret<Vec<u8>>, Secret<Vec<u8>>> for HmacSha512<N> {
    type ReturnType<T> = Result<T, error::ContainerError<error::CryptoError>>;

-    fn encode(&self, input: Vec<u8>) -> Self::ReturnType<Vec<u8>> {
+    fn encode(&self, input: Secret<Vec<u8>>) -> Self::ReturnType<Secret<Vec<u8>>> {
        let key = hmac::Key::new(ring::hmac::HMAC_SHA512, self.0.peek());

```

File: src/storage/db.rs [hunk 2]
```diff
@@ -345,23 +345,23 @@ impl super::FingerprintInterface for Storage {
     }
     async fn insert_fingerprint(
         &self,
-        card: types::CardNumber,
-        hash_key: Secret<String>,
+        data: Secret<String>,
+        key: Secret<String>,
     ) -> Result<types::Fingerprint, ContainerError<Self::Error>> {
-        let algo = HmacSha512::<1>::new(hash_key.expose().into_bytes().into());
+        let algo = HmacSha512::<1>::new(key.expose().into_bytes().into());
```
Comment (NishantJoshi00) [inline src/storage/db.rs:351]: Secret supports map operator

Comment (NishantJoshi00) [reply src/storage/db.rs:351]: https://github.com/juspay/hyperswitch/blob/main/crates/masking/src/secret.rs#L75

File: src/storage/db.rs [hunk 3]
```diff
@@ -345,23 +345,23 @@ impl super::FingerprintInterface for Storage {
...
 
-        let fingerprint_hash = algo.encode(card.into_bytes())?;
+        let fingerprint_hash = algo.encode(data.expose().into_bytes().into())?;
 
         let output = self
-            .find_by_fingerprint_hash(Secret::new(&fingerprint_hash))
+            .find_by_fingerprint_hash(fingerprint_hash.clone().expose().as_slice().into())
```
Comment (NishantJoshi00) [inline src/storage/db.rs:356]: Same here

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

File: src/storage/db.rs [hunk 4]
```diff
@@ -343,17 +343,17 @@ impl super::FingerprintInterface for Storage {
...
         key: Secret<String>,
     ) -> Result<types::Fingerprint, ContainerError<Self::Error>> {
-        let algo = HmacSha512::<1>::new(key.expose().into_bytes().into());
+        let algo = HmacSha512::<1>::new(key.map(|inner| inner.into_bytes()));
 
         let fingerprint_hash = algo.encode(data.expose().into_bytes().into())?;
 
         let output = self
-            .find_by_fingerprint_hash(fingerprint_hash.clone().expose().as_slice().into())
+            .find_by_fingerprint_hash(fingerprint_hash.clone().map(|inner| inner))
```
Comment (Chethan-rao) [inline src/storage/db.rs:356]:
```suggestion
            .find_by_fingerprint_hash(fingerprint_hash.clone())
```

**Review by Chethan-rao — CHANGES_REQUESTED** [review]

**Review by NishantJoshi00 — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Pull request merged**